crates/p2p-network/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+use std::{collections::HashSet, sync::Arc, time::Duration};
+
+use anyhow::{Context, Result};
+use gw_config::P2PNetworkConfig;
+use socket2::{SockRef, TcpKeepalive};
+use tentacle::{
+    async_trait,
+    builder::ServiceBuilder,
+    context::{ServiceContext, SessionContext},
+    multiaddr::{MultiAddr, Protocol},
+    secio::SecioKeyPair,
+    service::{
+        ProtocolMeta, Service, ServiceAsyncControl, ServiceError, ServiceEvent, TargetProtocol,
+    },
+    traits::{ProtocolSpawn, ServiceHandle},
+    ProtocolId, SubstreamReadPart,
+};
+
+const RECONNECT_DURATION: Duration = Duration::from_secs(5);
+
+/// Wrapper for tentacle Service. Automatcially reconnect dial addresses.
+pub struct P2PNetwork {
+    service: Service<SHandle>,
+}
+
+impl P2PNetwork {
+    pub async fn init<PS>(config: &P2PNetworkConfig, protocols: PS) -> Result<Self>
+    where
+        PS: IntoIterator,
+        PS::Item: Into<ProtocolMeta>,
+    {
+        #[allow(clippy::mutable_key_type)]
+        let mut dial = HashSet::with_capacity(config.dial.len());
+        for d in &config.dial {
+            let address: MultiAddr = d.parse().context("parse dial address")?;
+            dial.insert(address);
+        }
+        let dial_vec: Vec<MultiAddr> = dial.iter().cloned().collect();
+        let mut builder = ServiceBuilder::new()
+            .forever(true)
+            .tcp_config(|socket| {
+                let sock_ref = SockRef::from(&socket);
+                sock_ref.set_reuse_address(true)?;
```

Comment:
Setting the reuse address but not the bind address is useless for dial operations

You can refer to https://github.com/nervosnetwork/axon/blob/main/core/network/src/service.rs#L266

Reply:
This is for the listening socket. Without this there will often be “address already in use” error on `bind` due to TIME_WAIT sockets in local (i.e. not docker) testing.

Reply:
This setting affects all incoming and outgoing sockets, not just the listening socket

Reply:
Is there any harm for dial sockets?

Or how do I set it only for listening sockets?

Reply:
If it's just for testing, consider differentiating the ports from the configuration.

> Is there any harm for dial sockets

It's just one more operation that causes other processes to use the same port, otherwise there is no effect for now

Reply:
Are you referring to `SO_REUSEPORT`? `SO_REUSEADDR` does not allow other processes bind to the exact same (address, port).

Reply:
https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ

These two parameters do not behave the same on different platforms, on windows it is possible, on linux you have to set both reuse port and reuse address

Reply:
Binding to the same port for outgoing sockets seems rather strange to me. Why do that?

Anyway, I don't think this has any harm for our outgoing sockets.

Reply:
I thought at first that you were planning to use the reuse port feature for public network autodiscovery, but if not, there's no harm in setting it up this way